lib/bearerstrategy.js

```diff
@@ -118,121 +125,298 @@ const jws = require('jws');
-  // @TODO: this won't work - fetch is async, throwing in an async callback is not recommended because that exception can not be located.
-  // Also, this fetch would load parameters required by `BearerStrategy`. It is recommended to write own `authenticate` method for this Strategy
-  self.metadata.fetch((fetchMetadataError) => {
-    if (fetchMetadataError) {
-      throw new Error(`Unable to fetch metadata: ${fetchMetadataError}`);
+  jwt.verify(token, PEMkey, self._options, (err, verifiedToken) => {
+    if (err) {
+      if (err instanceof jwt.TokenExpiredError) {
+        log.warn('Access token expired');
+        return done(null, false, 'The access token expired');
+      }
+      if (err instanceof jwt.JsonWebTokenError) {
+        log.warn('An error was received validating the token', err.message);
+        return done(null, false, util.format('Invalid token (%s)', err.message));
+      }
+      return done(err, false);
     }
-
-    if (opts.validateIssuer) {
-      opts.issuer = self.metadata.oidc.issuer;
+    log.info(verifiedToken, 'was token going out of verification');
+    if (self._passReqToCallback) {
+      log.info('We did pass Req back to Callback');
+      return self._verify(req, verifiedToken, done);
     }
-    opts.algorithms = self.metadata.oidc.algorithms;
+    log.info('We did not pass Req back to Callback');
+    return self._verify(verifiedToken, done);
   });
+}
 
-  function jwtVerify(req, token, done) {
-    const decoded = jws.decode(token);
-    let PEMkey = null;
+Strategy.prototype.authenticate = function authenticateStrategy(req) {
+  const self = this;
 
-    if (decoded == null) {
-      done(null, false, 'Invalid JWT token.');
-    }
+  /* Some introduction to async.waterfall (from the following link):
+   * http://stackoverflow.com/questions/28908180/what-is-a-simple-implementation-of-async-waterfall
+   *
+   *   Runs the tasks array of functions in series, each passing their results 
+   * to the next in the array. However, if any of the tasks pass an error to 
+   * their own callback, the next function is not executed, and the main callback
+   * is immediately called with the error.
+   *
+   * Example:
+   *
+   * async.waterfall([
+   *   function(callback) {
+   *     callback(null, 'one', 'two');
+   *   },
+   *   function(arg1, arg2, callback) {
+   *     // arg1 now equals 'one' and arg2 now equals 'two'
+   *     callback(null, 'three');
+   *   },
+   *   function(arg1, callback) {
+   *     // arg1 now equals 'three'
+   *     callback(null, 'done');
+   *   }
+   * ], function (err, result) {
+   *      // result now equals 'done'    
+   * }); 
+   */
+  async.waterfall([
 
-    log.info('token decoded:  ', decoded);
+    // compute metadata url
+    (next) => {
+      if (!self._options.certificate && !self._options.identityMetadata) {
+        log.warn('No options was presented to Strategy as required.');
```

Comment:
I think this log message should be something like: "options.certificate && options.identityMetadata are both null. It is not possible to validate a JWT token. These values supply 'keys' required for validating signatures."